### DISCUSSION TITLE
Sourcing to global scope from function

### DISCUSSION THREAD

**Discussion by powercode** [PowerShell/PowerShell#14815] [Q&A]
Is it possible to write a function that reloads the profile?

That would require a function to manipulate the global scope. I have not found any mechanisms that do this. Am I missing something?

I've tried `Powershell.AddScript`.

This works if implemented in C#, where AddScript seems to run in the global scope, but when run in a function it only seems to change the local function scope.

test1.ps1 contains:
```powershell
$a = 1
```

Works:
```C#
    [Cmdlet(verbName: VerbsData.Import, "ScriptCmdlet")]
    public class ImportScriptCommand : PSCmdlet
    {
        protected override void EndProcessing()
        {
            using var ps = PowerShell.Create(RunspaceMode.CurrentRunspace);
            ps.AddScript(". ~/documents/powershell/test.ps1", false);
            var res = ps.Invoke();
            WriteObject(res, true);
        }
    }
```

Does not work:
```powershell
function Import-ScriptFunction {
    $ps = [PowerShell]::Create([RunspaceMode]::CurrentRunspace)
    $null = $ps.AddScript('. ~/documents/powershell/test.ps1', $false)
    $ps.Invoke()
    $ps.Dispose()
}
```

```powershell
Import-ScriptCmdlet
$a
```
```text
1
```

```powershell
Import-ScriptFunction
$a
```
```text
```

**Comment by powercode** [comment]
@BrucePay - this seems like something you would know :)

**Comment by ili101** [comment]
1. If it's Variable, Alias or Function you can set global scope https://github.com/PowerShell/PowerShell/issues/11178.
2. if it's something else you can actually dot source the function `. Import-ScriptCmdlet` but it's a bad solution as it rely on the user to "." your function.
3. If your function is in a module you can get your caller scope that may or may not be the global scope https://github.com/PowerShell/PowerShell/issues/6147

If someone finds something better I will like to know to.

**Comment by SeeminglyScience** [comment]
@powercode No supported way, but with a ton of reflection and stars aligning you can.

<details>

<summary>Example code</summary>

```powershell
$module = [psmoduleinfo]::new{
    function Invoke-Profile {
        [CmdletBinding()]
        param()
        end {
            $flags = [System.Reflection.BindingFlags]::NonPublic -bor
                [System.Reflection.BindingFlags]::Instance

            $context = $ExecutionContext.GetType().GetField('_context', $flags).GetValue($ExecutionContext)
            $ssi = $context.GetType().GetProperty('TopLevelSessionState', $flags).GetValue($context)
            $currentScope = $ssi.GetType().GetProperty('CurrentScope', $flags).GetValue($ssi)
            $globalScope = $ssi.GetType().GetProperty('GlobalScope', $flags).GetValue($ssi)

            if ($currentScope -ne $globalScope) {
                throw 'Top level SessionState currently has a child scope so this won''t work.'
            }


            $globalState = [psmoduleinfo]::new($false)
            $globalState.SessionState = $ssi.GetType().GetProperty('PublicSessionState', $flags).GetValue($ssi)

            # I don't remember the actual execution order, or if they all even get invoked
            # when multiple exist? I think so, but test first.
            . $globalState {
                if (Test-Path $profile.AllUsersAllHosts) {
                    . $profile.AllUsersAllHosts
                }

                if (Test-Path $profile.AllUsersCurrentHost) {
                    . $profile.AllUsersCurrentHost
                }

                if (Test-Path $profile.CurrentUserAllHosts) {
                    . $profile.CurrentUserAllHosts
                }

                if (Test-Path $profile.CurrentUserCurrentHost) {
                    . $profile.CurrentUserCurrentHost
                }
            }
        }
    }

    Export-ModuleMember -Function Invoke-Profile
}

Import-Module -ModuleInfo $module -Force
```
</details>

(echoing some of what @ili101 said, that's all correct)

- Every module has it's own tree of scopes called a `SessionState`
- Global has it's own `SessionState`
- You can invoke a scriptblock in the context of a different `SessionState`
- Using the "dot source" operator doesn't bring state into the current session state, it just doesn't branch the tree it's assigned to
- Since you aren't **really** invoking in top level `SessionStateScope` directly, this won't work if there's any currently executing child scopes in the global `SessionState`

You could also skip all that, load that function in global, and just dot source it like `. Invoke-Profile` but I'm guessing that's not what you're lookin' for. Otherwise nah, once the scope is created you can't unmake it from the function running in it.

**Disclaimer:** Probably just don't tho
